mount/mount_darwin.go

```diff
@@ -0,0 +1,59 @@
+//go:build darwin
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package mount
+
+import (
+	"fmt"
+	"os/exec"
+)
+
+// Mount to the provided target.
+func (m *Mount) mount(target string) error {
+	var commandName string
+	if m.Type == "bind" {
+		// macOS doesn't natively support bindfs/nullfs
+		// The way to emulate it is via FUSE fs named "bindfs"
+		commandName = "bindfs"
```

Comment:
`bindfs` is a third party package that depends on a proprietary FUSE on macOS, so I don't think we want to introduce a dependency on this.

Can we just make this a third party snapshotter plugin?

Reply:
No we can't because mounting is not only used through snapshotter API. There are multiple places throughout containerd that want to mount and unmount stuff. See `runtime/v2/bundle.go` in this PR, for example. Or `diff/apply/apply_other.go`. Those places have zero relationship with snapshotter API.

Reply:
I haven't looked through https://github.com/containerd/containerd/pull/8807 yet, possibly it adds some kind of abstraction that allows mount/unmount to be truly pluggable. But the very fact that they PR tons of stuff into containerd instead of making an "external snapshotter" hints that no, the abstraction layer is too leaky to keep that outside of containerd.

Reply:
I'd still prefer https://github.com/containerd/containerd/pull/5935 over this PR for pluggability

Reply:
#5935 is waiting for two years already, I thought there are some issues with it. And there are: #5935 doesn't actually allow different mount mechanisms - how one is supposed to put more than one `mount_containerd_darwin` on PATH?

BTW, I looked through https://github.com/containerd/containerd/pull/8807. It does absolutely nothing to `mount/mount_windows.go`, it is hardcoded to use hcsshim and `windows-layer`.

So, while I agree with you that mount mechanisms should be 1. pluggable 2. tied to snapshotter, this is not how things currently work in containerd.

Reply:
I'd like to avoid mixing "fix snapshotter API to actually abstract away underlying stuff from containerd" with "add _any_ working snapshotter to macOS". Because the latter severely increases the scope of the task, involves nontrivial design decisions and better be done by someone with high expertise of containerd internals.

Reply:
Or, can we just pretend that I did what you suggest and `bindfs` is _the name of external snapshotter plugin_?) Will things be better if I rename `bindfs` in this PR. to`containerd-macos-mount-helper` while leaving everything else as-is?

Reply:
@AkihiroSuda As you wish!

1. containerd now calls an absolutely neutral `containerd-darwin-mount-helper` binary whose CLI is suspiciously similar to `mount(8)` (see PR update)
2. There is [a reference implementation of `containerd-darwin-mount-helper`](https://github.com/slonopotamus/containerd-darwin-mount-helper) that accidentally happens to be just a wrapper around `bindfs`

Just in case: when someone comes across these commits in the future, I have absolutely zero objection against throwing all this out and replacing with a proper abstraction! We just didn't have one yet.

Reply:
Now resulting delta of this PR is _minus 14 lines_! Did you ever implement a feature writing negative amount of code? :D

Reply:
> Did you ever implement a feature writing negative amount of code? :D

Can confidently say no 🤣